### PR TITLE
[ROCm] init hipblas_lt in XLA

### PR DESCRIPTION
**Pull request by i-chaochen**
The original one from https://github.com/tensorflow/tensorflow/pull/60894

We made some changes based on the latest upstream changes and split this big one to XLA first.

@akuegel your previous comments have been resolved, please check. Thanks a lot for the reviewing!

### CODE REVIEW

File: xla/debug_options_flags.cc
Comment (akuegel) [inline xla/debug_options_flags.cc:84]: I guess this was not an intentional change (probably merge conflict?). Please revert the change in this file.

File: xla/service/gpu/cublas_lt_matmul_thunk.cc
Comment (akuegel) [inline xla/service/gpu/cublas_lt_matmul_thunk.cc:18]: Could this be handled in the build file instead?
So not use if_gpu_is_configured but if_cuda_is_configured and if_rocm_hipblaslt, same as for ir_emitter_triton.

File: xla/service/gpu/gpu_compiler.cc
Comment (akuegel) [inline xla/service/gpu/gpu_compiler.cc:490]: Can you please create virtual method for this, and implement it differently in amdgpu_compiler and nvptx_compiler? Similar to how it was done in https://github.com/openxla/xla/commit/eff2b737cb4f6949fa85e3646eed3db684826dfc

File: xla/service/gpu/runtime/BUILD
Comment (akuegel) [inline xla/service/gpu/runtime/BUILD:431]: This dependency needs to be behind a guard.

File: xla/service/gpu/runtime/cublas_lt_matmul.cc
Comment (akuegel) [inline xla/service/gpu/runtime/cublas_lt_matmul.cc:35]: This include needs to be behind a guard.

File: xla/service/gpu/runtime/cublas_lt_matmul.h
Comment (akuegel) [inline xla/service/gpu/runtime/cublas_lt_matmul.h:22]: This include needs to be behind a guard.

File: xla/service/gpu/matmul_utils.h
Comment (akuegel) [inline xla/service/gpu/matmul_utils.h:36]: Why does it have #if GOOGLE_CUDA here?
I think the include is only needed for ROCM?

File: xla/stream_executor/rocm/BUILD
Comment (akuegel) [inline xla/stream_executor/rocm/BUILD:485]: Please put the hdrs and deps behind if_rocm_is_configured guards.

File: xla/stream_executor/rocm/BUILD
Comment (akuegel) [inline xla/stream_executor/rocm/BUILD:458]: Please put the hdrs, srcs and deps behind if_rocm_is_configured guards.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/debug_options_flags.cc
Comment (i-chaochen) [reply xla/debug_options_flags.cc:84]: Yes, I think this is rollback to 0 now. https://github.com/openxla/xla/pull/3959

File: xla/service/gpu/cublas_lt_matmul_thunk.cc
Comment (i-chaochen) [reply xla/service/gpu/cublas_lt_matmul_thunk.cc:18]: @akuegel I am sorry I am not sure I am 100% understand this one "be handled in the build file" do you mean to remove this preprocessor #if GOOGLE_CUDA || TF_HIPBLASLT in cubas_lt_matmul_thunk.cc ?  

the BUILD file for ```cublas_lt_matmul_thunk.cc``` and ```ir_emitter_triton.cc``` are using ```if_cuda_is_configured``` and ```if_rocm_hipblaslt```. 

https://github.com/openxla/xla/pull/3953/files#diff-35357ad3b3b70062889668b03399a71db3249cd2a01e90e0b0ed8ebe9c817272R1257-R1263

https://github.com/openxla/xla/pull/3953/files#diff-35357ad3b3b70062889668b03399a71db3249cd2a01e90e0b0ed8ebe9c817272R416-R417

File: xla/service/gpu/gpu_compiler.cc
Comment (i-chaochen) [reply xla/service/gpu/gpu_compiler.cc:490]: Much apprecipate for the example. I will look it.

File: xla/service/gpu/cublas_lt_matmul_thunk.cc
Comment (akuegel) [reply xla/service/gpu/cublas_lt_matmul_thunk.cc:18]: What I meant is: for triton_emitter target, your PR changes it to:

```
srcs = if_cuda_is_configured(["ir_emitter_triton.cc"])
        + if_rocm_hipblaslt(["ir_emitter_triton.cc"]),
```

I thought that the same might be possible for the cublas_lt_matmul_thunk target. Then, I think the preprocessor #if may not be needed.

File: xla/stream_executor/rocm/BUILD
Comment (i-chaochen) [reply xla/stream_executor/rocm/BUILD:458]: Done

File: xla/service/gpu/cublas_lt_matmul_thunk.cc
Comment (i-chaochen) [reply xla/service/gpu/cublas_lt_matmul_thunk.cc:18]: Got it! Grateful for your patience in explaining!

File: xla/service/gpu/gpu_compiler.cc
Comment (i-chaochen) [reply xla/service/gpu/gpu_compiler.cc:490]: Just want to make sure in case I misunderstood here. Do you expect to have two functions somehow like this:

```
    TF_RETURN_IF_ERROR(AddOperandUpcasterPasses(
      &pipeline, hlo_module, stream_exec, debug_options, options,
      gpu_target_config, upcaster_filter));

    TF_RETURN_IF_ERROR(AddResultCasterPasses(
      &pipeline, hlo_module, stream_exec, debug_options, options,
      gpu_target_config, upcaster_filter));
```

Then, these two virtual functions will be implemented differently in amdgpu_compiler and nvptx_compiler, respectively. The only difference I think might just this one:
```
upcaster_filter = [&](const HloInstruction* instr) {
      return !stream_exec->GetDeviceDescription()
                  .cuda_compute_capability()
                  .IsAtLeast(se::CudaComputeCapability::VOLTA) ||
             !gpu::IsMatrixMultiplication(*instr);
}
```

If the purpose of doing the above is to remove the annoying preprocessor #if #else things, I can use gpu_target_config.platform_name or gpu_version to determine it at the runtime.

Like this 
```
      if (gpu_target_config.platform_name == "ROCM"){
        return !gpu::IsMatrixMultiplication(*instr);
      }
```

Comment (akuegel) [reply xla/service/gpu/gpu_compiler.cc:490]: Thanks, determining it at runtime is much better :)

Comment (i-chaochen) [reply xla/service/gpu/gpu_compiler.cc]: Done.

File: xla/service/gpu/gpu_compiler.cc
Comment (akuegel) [inline xla/service/gpu/gpu_compiler.cc:458]: nit: space between "else" and "{"
also should go to the previous line:

```
} else {
```

**Review by akuegel — APPROVED** [review]

File: xla/stream_executor/rocm/hip_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/rocm/hip_blas_lt.cc:17]: This include is not ok. It is only allowed to include header files.

File: xla/stream_executor/rocm/rocm_blas.h
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_blas.h:102]: Please add a blank line in front of private:

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_blas_lt.cc:280]: Please use static_cast<int>(error) instead of int(error)

File: xla/stream_executor/rocm/rocm_helpers.cu.cc
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_helpers.cu.cc:94]: Please use static_cast<float>() instead of float()

File: xla/stream_executor/rocm/rocm_helpers.cu.cc
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_helpers.cu.cc:121]: Please use static_cast<uint64_t>() instead of uint64_t()

File: xla/stream_executor/rocm/rocm_helpers.cu.cc
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_helpers.cu.cc:122]: Please use static_cast<uint64_t>() instead of uint64_t()

File: xla/service/gpu/matmul_utils.h
Comment (akuegel) [inline xla/service/gpu/matmul_utils.h:44]: Can you use #elif here?
Just to make it clear that we don't have a duplicate include.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/service/gpu/matmul_utils.h
Comment (i-chaochen) [reply xla/service/gpu/matmul_utils.h]: Done. Thanks a lot for the check.

https://github.com/openxla/xla/pull/3953/files#diff-98376f6c713755983f4458b3660033af9384bf588a8aea20e226624d7fcefeeeR55-R59

It seems ```#if GOOGLE_CUDA``` doesn't work in cuda_blas_lt.cc and failed to compile at cuda side, I am not familar with cuda side, wonder do you have any idea why it doesn't work there? Thanks in advance!

File: xla/stream_executor/rocm/hip_blas_lt.cc
Comment (akuegel) [reply xla/stream_executor/rocm/hip_blas_lt.cc:17]: This doesn't seem resolved yet. The linter complains that you are including a .cc file. It needs to be a .h file.

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (akuegel) [reply xla/stream_executor/cuda/cuda_blas_lt.cc:278]: This doesn't look resolved yet.

File: xla/stream_executor/cuda/BUILD
Comment (akuegel) [inline xla/stream_executor/cuda/BUILD:258]: Please do not put .cc files as headers. If it is a header, it should be a .h file.

File: xla/service/gpu/matmul_utils.h
Comment (akuegel) [reply xla/service/gpu/matmul_utils.h]: It is possible that GOOGLE_CUDA is not defined. Ideally, we would not need GOOGLE_CUDA macros inside the stream_exector/cuda directory. So far, we only need it in tests. Can you maybe find a different way to share code and create a shared target outside the cuda directory? I have not thought about this myself, so maybe it is not possible. Your current usage of cuda_blas_lt.cc as header is definitely not ok though.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/stream_executor/rocm/hip_blas_lt.cc
Comment (i-chaochen) [reply xla/stream_executor/rocm/hip_blas_lt.cc]: Sorry, I forgot to add this.

File: xla/stream_executor/rocm/hip_blas_lt.h
Comment (akuegel) [inline xla/stream_executor/rocm/hip_blas_lt.h:80]: Please use static_cast<int> instead of (int)

File: xla/stream_executor/rocm/rocm_dnn.cc
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_dnn.cc:3965]: Please use static_cast<int> instead of (int)

File: xla/stream_executor/rocm/rocm_dnn.cc
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_dnn.cc:3939]: Please use static_cast here and in the two lines below

File: xla/stream_executor/rocm/rocm_dnn.cc
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_dnn.cc:3910]: Please use static_cast here and in the lines below

File: xla/stream_executor/rocm/rocm_helpers.cu.cc
Comment (akuegel) [inline xla/stream_executor/rocm/rocm_helpers.cu.cc:147]: Please use static_cast for casting in the two cases in this line

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/stream_executor/cuda/cuda_blas_lt.h
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_blas_lt.h:37]: To me this looks like an indication that it would be better to have this file one directory up, and name it gpu_blas_lt.h
Also, the implementation file likely should be one directory up, and cuda and rocm implementation could be subclasses that have specific virtual methods for where they differ.

**Review by akuegel — CHANGES_REQUESTED** [review]
I was out of office on Thursday and Friday, sorry for letting you wait for another review.

File: xla/stream_executor/cuda/cuda_blas_lt.h
Comment (i-chaochen) [reply xla/stream_executor/cuda/cuda_blas_lt.h:37]: Yes, it's much better in this way.

Comment (i-chaochen) [reply xla/stream_executor/cuda/cuda_blas_lt.h:37]: Hi @akuegel I have fixed CI checks on cuda side, wonder would be possible to get merged first and I will send another PR to redesign this part? as this hipblas_lt is urgent and important to us due to JAX.

As a side note, Check Contents / check-contents keeps failed for tsl::Status, but I can see other merged PRs are ok with tsl::Status(), e.g., https://github.com/openxla/xla/commit/386fd8513ea7430a527e4917e28fb6f1262c0d4b

File: third_party/llvm/workspace.bzl
Comment (akuegel) [inline third_party/llvm/workspace.bzl:7]: This should not be changed. This always needs to match with the version we use internally at Google.

File: third_party/tsl/third_party/llvm/workspace.bzl
Comment (akuegel) [inline third_party/tsl/third_party/llvm/workspace.bzl:7]: Same here, this should not be changed.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/stream_executor/cuda/cuda_blas_lt.h
Comment (akuegel) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: I would be ok if you just copy the code for now, and later work on unifying it with a shared base class. But please don't add rocm specific code inside the cuda directory.

File: third_party/llvm/workspace.bzl
Comment (i-chaochen) [reply third_party/llvm/workspace.bzl]: Thanks for pointing it out. Yes, this is irrelevent one. I have cherry-pick it back.

File: third_party/tsl/third_party/llvm/workspace.bzl
Comment (i-chaochen) [reply third_party/tsl/third_party/llvm/workspace.bzl]: Thanks for pointing it out. Yes, this is irrelevent one. I have cherry-pick it back.

File: xla/stream_executor/cuda/cuda_blas_lt.h
Comment (i-chaochen) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: Thanks for the advice. Initially we think this is the leastest code change to achieve....Yes, we wouldn't do it again.

Comment (i-chaochen) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: Hi @ddunl wonder will be ok to merge this PR first? as it's urgent and important to use having JAX to use hipblas_lt.

@akuegel is ok for now and I will create another PR (it's on progress) to redesign and unify hipblas_lt and cublas_lt.

Comment (ddunl) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: I think I'd have trouble getting this merged with the rocm references under stream_executor/cuda, but if you can fix that and have tests passing I can try to merge it later today

Comment (i-chaochen) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: Thanks a lot! @ddunl  

It's already passed all tests in CI and it's #ifndef TENSORFLOW_USE_ROCM, so it wouldn't call any rocm references in cuda side.

Comment (ddunl) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: I mean even if it's ifdef'd out, I should be able to do `grep -R rocm` under `stream_executor/cuda` and get 0 results, that's what I think I'll have trouble merging if I import it

Comment (i-chaochen) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: Ok, I will try to copy past and move it out first. Thanks for clarification!

Comment (akuegel) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: Thanks @ddunl for explaining what I meant :)
Yes, we also don't want any ROCM specific #ifdefs in stream_executor/cuda. Copying the code and adjusting it is also not great, but still better. And then you can work on unifying the code with a common base class later in a separate PR.

Comment (ekuznetsov139) [reply xla/stream_executor/cuda/cuda_blas_lt.h]: Would it satisfy you if we moved the existing cuda_blas_lt.cc entirely into stream_executor/gpu instead?

**Comment by i-chaochen** [comment]
I have moved the rocm part from cuda directory to our rocm directory. Hope now it's ok.

@akuegel

**Review by akuegel — APPROVED** [review]
Thank you :)

**Comment by i-chaochen** [comment]
@ddunl akuegel is approved, wonder will this be merged today? Thanks in advance!

File: xla/stream_executor/rocm/hip_blas_lt.cc
Comment (ddunl) [inline xla/stream_executor/rocm/hip_blas_lt.cc:31]: Seems like this is already included on line 26

**Comment by ddunl** [comment]
Just one comment, but otherwise we are experiencing infra issues that should be resolved today so we can merge it, but can't guarantee that. I'll try to get it merged today

**Comment by i-chaochen** [comment]
> Just one comment, but otherwise we are experiencing infra issues that should be resolved today so we can merge it, but can't guarantee that. I'll try to get it merged today

Thanks! Just a side note, I notice there is no gpu test anymore on this PR, do you know why?

**Comment by ddunl** [comment]
That's a symptom of the infra issues I was referring to - there should be, but it probably timed out waiting for a GPU last time it tried to run. When you delete the extra include and push it should run again and (hopefully) run the GPU tests

File: xla/stream_executor/rocm/hip_blas_lt.cc
Comment (i-chaochen) [reply xla/stream_executor/rocm/hip_blas_lt.cc:30]: Done

**Comment by ddunl** [comment]
Oh sorry looks like there's conflicts now, can you rebase?

**Comment by i-chaochen** [comment]
> Oh sorry looks like there's conflicts now, can you rebase?

Done!

**Comment by ddunl** [comment]
Looks good, running internal tests now

**Comment by i-chaochen** [comment]
@ddunl hmmm....it seems that has conflicts during copybara-service is merging...should I resolve it?

**Comment by ddunl** [comment]
I think we are ok, my patch just rebased to HEAD internally without issue. Hopefully should get merged today

**Comment by ddunl** [comment]
Possible that there's an internal test failure on this one but not sure, looking into it more

**Comment by akuegel** [comment]
Just in case you are not aware, your change doesn't correctly build if hipblas build is not available. I am still trying to fix it, but might not get it done today. You can see my progress here:

https://github.com/openxla/xla/pull/4406

**Pull request closed without merging**
